lib/floe/workflow/intrinsic_function/transformer.rb

```diff
@@ -3,41 +3,189 @@
 # rubocop:disable Performance/RegexpMatch, Performance/RedundantMatch
 
 require "parslet"
-require "securerandom"
 
 module Floe
   class Workflow
     class IntrinsicFunction
       class Transformer < Parslet::Transform
-        def self.resolve_args(args)
-          if args.nil?
-            # 0 args
-            []
-          elsif args.kind_of?(Array)
-            # >1 arg
-            args.map { |a| a[:arg] }
-          else
-            # 1 arg
-            [args[:arg]]
+        OptionalArg = Struct.new(:type)
+
+        class << self
+          def process_args(args, function, signature = nil)
+            args = resolve_args(args)
+            if signature
+              check_arity(args, function, signature)
+              check_types(args, function, signature)
+            end
+            args
+          end
+
+          private
+
+          def resolve_args(args)
+            if args.nil?
+              # 0 args
+              []
+            elsif args.kind_of?(Array)
+              # >1 arg
+              args.map { |a| a[:arg] }
+            else
+              # 1 arg
+              [args[:arg]]
+            end
+          end
+
+          def check_arity(args, function, signature)
+            if signature.any?(OptionalArg)
+              signature_without_optional = signature.reject { |a| a.kind_of?(OptionalArg) }
+              signature_size = (signature_without_optional.size..signature.size)
+
+              raise ArgumentError, "wrong number of arguments to #{function} (given #{args.size}, expected #{signature_size})" unless signature_size.include?(args.size)
+            else
+              raise ArgumentError, "wrong number of arguments to #{function} (given #{args.size}, expected #{signature.size})" unless signature.size == args.size
+            end
+          end
+
+          def check_types(args, function, signature)
+            args.zip(signature).each_with_index do |(arg, type), index|
+              type = type.type if type.kind_of?(OptionalArg)
+
+              raise ArgumentError, "wrong type for argument #{index + 1} to #{function} (given #{arg.class}, expected #{type})" unless arg.kind_of?(type)
+            end
           end
         end
 
         rule(:null_literal  => simple(:v)) { nil }
         rule(:true_literal  => simple(:v)) { true }
         rule(:false_literal => simple(:v)) { false }
 
-        rule(:string   => simple(:v)) { v.to_s }
+        rule(:string   => simple(:v)) { v.to_s[1..-2] }
```

Comment:
I'm assuming this is to extract `foo` from `"foo"` ?

Reply:
Yes, so I had to make a change because empty string wasn't being handled properly.  More details in the commit message here: https://github.com/ManageIQ/floe/pull/242/commits/3a3640650f8521e899d7ff3184b7554d69aacc8d

>  Fix issue where empty string would not parse as expected
>
> Normal string would parse into a slice, however an empty string would
> parse into an empty array. Thus the transformer would not match because
> simple can only match on scalar values. Instead, we now include the
> quotes in the string and remove them afterwards.